cli/tools/registry/pm.rs

```diff
@@ -25,22 +28,149 @@ use crate::file_fetcher::FileFetcher;
 use crate::jsr::JsrFetchResolver;
 use crate::npm::NpmFetchResolver;
 
+enum ConfigFile {
+  Deno(deno_config::ConfigFile),
+  Npm(deno_node::PackageJson, Option<FmtOptionsConfig>),
+}
+
+impl ConfigFile {
+  fn specifier(&self) -> ModuleSpecifier {
+    match self {
+      Self::Deno(d) => d.specifier.clone(),
+      Self::Npm(n, ..) => n.specifier(),
+    }
+  }
+
+  /// Returns the existing imports/dependencies from the config.
+  fn existing_imports(&self) -> Result<IndexMap<String, String>, AnyError> {
+    match self {
+      ConfigFile::Deno(deno) => {
+        if let Some(imports) = deno.json.imports.clone() {
+          match serde_json::from_value(imports) {
+            Ok(map) => Ok(map),
+            Err(err) => {
+              bail!("Malformed \"imports\" configuration: {err}")
+            }
+          }
+        } else {
+          Ok(Default::default())
+        }
+      }
+      ConfigFile::Npm(npm, ..) => {
+        Ok(npm.dependencies.clone().unwrap_or_default())
+      }
+    }
+  }
+
+  fn fmt_options(&self) -> FmtOptionsConfig {
+    match self {
+      ConfigFile::Deno(deno) => deno
+        .to_fmt_config()
+        .ok()
+        .flatten()
+        .map(|f| f.options)
+        .unwrap_or_default(),
+      ConfigFile::Npm(_, config) => config.clone().unwrap_or_default(),
+    }
+  }
+
+  fn imports_key(&self) -> &'static str {
+    match self {
+      ConfigFile::Deno(_) => "imports",
+      ConfigFile::Npm(..) => "dependencies",
+    }
+  }
+
+  fn file_name(&self) -> &'static str {
+    match self {
+      ConfigFile::Deno(_) => "deno.json",
+      ConfigFile::Npm(..) => "package.json",
+    }
+  }
+
+  fn is_npm(&self) -> bool {
+    matches!(self, Self::Npm(..))
+  }
+
+  /// Get the preferred config file to operate on
+  /// given the flags. If no config file is present,
+  /// creates a `deno.json` file - in this case
+  /// we also return a new `CliFactory` that knows about
+  /// the new config
+  async fn from_flags(flags: Flags) -> Result<(Self, CliFactory), AnyError> {
+    let factory = CliFactory::from_flags(flags.clone())?;
+    let options = factory.cli_options().clone();
+
+    match (options.maybe_config_file(), options.maybe_package_json()) {
+      // when both are present, for now,
+      // default to deno.json
+      (Some(deno), Some(_) | None) => {
+        Ok((ConfigFile::Deno(deno.clone()), factory))
+      }
+      (None, Some(package_json)) if options.enable_future_features() => {
+        Ok((ConfigFile::Npm(package_json.clone(), None), factory))
+      }
+      (None, Some(_) | None) => {
+        tokio::fs::write(options.initial_cwd().join("deno.json"), "{}\n")
```

Comment:
It's probably preferable to not use tokio for fs operations outside the JS runtime code because there's not really any reason to and then this doesn't need to be an async method.